### ISSUE TITLE
Support sounder driver

### ISSUE THREAD

**Issue by leskerr** [MorseKOB/PyKOB#111]
MKOB currently works only with a loop interface to an external key and sounder. An option needs to be added to allow for a sounder driver interface (or even just a key connected directly to the serial port and no sounder).

These options are handled very poorly in MorseKOB 2.5 and the result has proved to be very confusing (even to its developer!) This change needs to be carefully thought out.

**Comment by AESilky** [comment]
Are the (to be) supported combinations:
1. No external equipment - Keyboard, Computer Audio
2. External KOB - Loop Interface (Computer Audio off)
3. External key, external sounder (separate) - Key Interface, Sounder driver
4. External key, no sounder - Key Interface, Computer Audio
5. No key, external sounder - Keyboard, Sounder driver
Is that correct?

How much different is supporting the Loop Interface (KOB) and supporting the Key Interface + Sounder Driver?

**Comment by pwdirks** [comment]
Are these combinations best described as a choice of input (keyboard, key in external KOB loop interface, direct external key interface) and a bunch of separate outputs (in addition to the screen display), some or all of which may be enabled (computer audio, external KOB loop interface)?  Is there another output option for some external interface?

**Comment by AESilky** [comment]
External sounder (not wired together with the key (KOB)).

I believe that it needs to know if it is a KOB or a separate key and sounder. @leskerr please correct me if I'm wrong.

Though it is technically possible to enable both the simulated sounder and a 'real' external sounder (either stand-alone or in a KOB using the loop interface), I'm not sure why you would do that. It seems likely to me that they could be slightly out of sync, and the sound/tone would certainly be different between the two.

I use the simulated sounder when I'm working without my KOB connected or if I'm working late and I want to turn down the volume or put on headphones so I don't disturb others.

I would like to be able to use my KOB key (using the loop interface) and the simulated sounder. My reason is that I like the key on my KOB better than my stand-alone key, but if I'm training/playing/testing late at night (which is most of the time) I need to use the simulated sounder so I can use headphones.

**Comment by leskerr** [comment]
The essential aspect of supporting the sounder driver interface is that inputs received from the key need to be echoed back to the sounder. `pykob/kob.py` already provides for this functionality with the `echo` parameter.

I'm not sure the code associated with that option has ever been tested. I haven't been able to find my sounder driver hardware, and that's an impediment to testing right now.

I never anticipated a need for using a loop interface with the sounder in the loop being silenced. Now that I think about it, I can see how it could be useful. It's a feature I'd appreciate using myself at times.

**Comment by leskerr** [comment]
I think we need a new config option, `Interface`, with possible values of `Key & Sounder`, `Loop`, and (for the future) `Keyer`.

**Comment by AESilky** [comment]
Issue #147 added for the config option mentioned by @leskerr .

**Comment by leskerr** [comment]
I used the interface options provided by @AESilky to implement the 'Key & Sounder' option in MKOB. I started testing with a key and speakers instead of an external sounder. (I don't have the capability to test the sounder driver yet.)

The changes work to some extent, but there's a significant delay between opening or closing the key and the sound heard from the speaker. I recorded the audio, which captures the click of the key opening or closing and the corresponding sound from the speaker. I measured the delay by analyzing the audio waveform and found that it ranged from 118 ms to 138 ms (in 12 trials). This test used PyAudio, not sounddevice.

I don't know where this latency is coming from or how to fix it, but it makes it quite difficult to send on a bug.
